lib/sycamore/sycamore/transforms/aggregation.py

```diff
@@ -0,0 +1,259 @@
+                    return accumulator
+                extra_metadata: list[MetadataDocument] = []
+                with ThreadLocal(ADD_METADATA_TO_OUTPUT, extra_metadata):
+                    final_doc = self._syc_agg._finalize(doc)
+                final_doc["key"] = accumulator["key"]
+                meta.extend(update_lineage(from_docs=[doc], to_docs=[final_doc]))
+                meta.extend(extra_metadata)
+                return {"doc": final_doc.serialize(), "meta": pickle.dumps(meta)}
+
+        ray_agg = RayAggregation(self, self._name, self._zero_factory)
+        ds = dataset.map(self._to_key_val).groupby("key").aggregate(ray_agg).flat_map(self._unpack)
+        return ds
+
+    def local_execute(self, all_docs: list[Document], do_combine: bool = True) -> list[Document]:
+        import random
+
+        documents, metadata = split_data_metadata(all_docs)
+        split_docs: dict[str, list] = {}
+
+        for d in documents:
+            key = self._group_key_fn(d)
+            if (split := split_docs.get(key, None)) is not None:
+                split.append(d)
+            else:
+                split_docs[key] = [d]
+
+        ret: list[Union[Document, MetadataDocument]] = metadata  # type: ignore
+        extra_metadata: list[MetadataDocument] = []
+        with ThreadLocal(ADD_METADATA_TO_OUTPUT, extra_metadata):
+            for key, split in split_docs.items():
+                if do_combine and len(split) > 1:
+                    # This path is mostly for testing. Otherwise combine is not exercised
+                    # in local mode. We also combine in a random order because if
+                    # we're testing we're trying to break your assumptions.
+                    documents_beginning = split[: len(split) // 2]
+                    documents_end = split[len(split) // 2 :]
+                    partial_beginning = self._accumulate(documents_beginning)
+                    ret.extend(update_lineage(from_docs=documents_beginning, to_docs=[partial_beginning]))
+                    partial_end = self._accumulate(documents_end)
+                    ret.extend(update_lineage(from_docs=documents_end, to_docs=[partial_end]))
+                    if random.random() < 0.5:
+                        partial_beginning, partial_end = partial_end, partial_beginning
+                    partial = self._combine(partial_beginning, partial_end)
+                    ret.extend(update_lineage(from_docs=[partial_beginning, partial_end], to_docs=[partial]))
+                else:
+                    partial = self._accumulate(split)
+                    ret.extend(update_lineage(from_docs=split, to_docs=[partial]))
+                final = self._finalize(partial)
+                ret.extend(update_lineage(from_docs=[partial], to_docs=[final]))
+                ret.append(final)
+        ret.extend(extra_metadata)
+        return ret
+
+
+class Aggregation:
+    def __init__(
+        self,
```

Comment:
I think the common case is going to be people inheriting from Aggregation, so I'd make that case easier.
Also I wouldn't use _ for the interface since it's effectively public.
```
name: Optional[str] = None,
accumulate_docs: Optional[Callable[list[Documnet]], Document] = None,
...
self._name = name or self.__class__.__name__
if accumulate_docs is not None:
    self.accumulate = lambda s, ds: accumulate_docs(ds)
...
def accumulate(self, in: list[Document]) -> Document:
    raise UnimplementedError("Must define accumulate function")
...
def finalize(self, d: Document) -> Document:
    return d

def zero_factory(self) -> Document:
    return Document()
```